orttraining/orttraining/core/framework/module_gradient_graph_builder.cc

```diff
@@ -179,46 +179,58 @@ void ModuleGradientGraphBuilder::AddYieldOp() {
     user_output_grad_names_set.insert(name + "_grad");
   }
 
-  // If an NodeArg is output of one of nodes, it's not the user output gradient needed by backward graph.
-  std::unordered_set<std::string> non_backward_user_output_grad_names;
+  // If an output gradient is output of one of nodes, need to add this output to PT's output gradient.
+  std::unordered_set<std::string> internal_output_grad_names;
   for (auto node_index : gradient_node_topology_list) {
     auto& node = *gradient_graph.GetNode(node_index);
     for (const auto& node_arg : node.OutputDefs()) {
       if (user_output_grad_names_set.find(node_arg->Name()) != user_output_grad_names_set.end()) {
-        non_backward_user_output_grad_names.insert(node_arg->Name());
+        internal_output_grad_names.insert(node_arg->Name());
       }
     }
   }
 
-  // YieldOps required_grad attribute specifies the indices of the required gradients.
-  ONNX_NAMESPACE::AttributeProto required_grad;
-  const std::string attribute_name = "required_grad";
-  required_grad.set_name(attribute_name);
-  required_grad.set_type(ONNX_NAMESPACE::AttributeProto::INTS);
-
-  training_graph_info_.backward_output_grad_names_map.clear();
-  for (std::size_t i = 0; i < training_graph_info_.user_output_names.size(); ++i) {
-    const auto& name = training_graph_info_.user_output_names[i];
-    std::string grad_name = name + "_grad";
-    if (non_backward_user_output_grad_names.find(grad_name) == non_backward_user_output_grad_names.end()) {
-      training_graph_info_.backward_output_grad_names_map.insert(std::make_pair(grad_name, i));
-      required_grad.add_ints(static_cast<int64_t>(i));
+  for (const auto& output_grad_name : internal_output_grad_names) {
+    Node* producer_node = gradient_graph.GetMutableProducerNode(output_grad_name);
+    NodeArg* arg = gradient_graph.GetNodeArg(output_grad_name);
+    int producer_node_arg_index = 0;
```

Comment:
In graph_utils.h, there is a GetNodeInputIndexFromInputName

maybe add a GetNodeOutputIndexFromOutputName ?